src/elastio-snap.c

```diff
@@ -2271,13 +2537,173 @@ static int cow_sync_and_close(struct cow_manager *cm){
 	return ret;
 }
 
+static inline void elastio_snap_mm_lock(struct mm_struct *mm)
+{
+	// TODO: fix for other kernel versions
+	down_write(&mm->mmap_sem);
+}
+
+static inline void elastio_snap_mm_unlock(struct mm_struct *mm)
+{
+	// TODO: fix for other kernel versions
+	up_write(&mm->mmap_sem);
+}
+
+static int elastio_snap_get_cow_file_extents(struct snap_device *dev, struct file *filp)
+{
+	int ret;
+	struct fiemap_extent_info fiemap_info;
+	unsigned int fiemap_mapped_extents_size, i_ext;
+	struct fiemap_extent *extent;
+	char parent_process_name[TASK_COMM_LEN];
+	unsigned long vm_flags = VM_READ | VM_WRITE;
+	unsigned long start_addr;
+	struct task_struct *task;
+	struct vm_area_struct *vma;
+	struct page *pg;
+	__user uint8_t *cow_ext_buf;
+	size_t cow_ext_buf_size;
+
+	int (*fiemap)(struct inode *, struct fiemap_extent_info *, u64 start, u64 len);
+
+	struct vm_area_struct * (*vm_area_alloc)(struct mm_struct *mm) = (VM_AREA_ALLOC_ADDR != 0) ?
+        (struct vm_area_struct * (*)(struct mm_struct *mm)) (VM_AREA_ALLOC_ADDR + (long long)(((void *)kfree) - (void *)KFREE_ADDR)) : NULL;
+
+	void (*vm_area_free)(struct vm_area_struct *vma) = (VM_AREA_FREE_ADDR != 0) ?
+        (void (*)(struct vm_area_struct *vma)) (VM_AREA_FREE_ADDR + (long long)(((void *)kfree) - (void *)KFREE_ADDR)) : NULL;
+
+	int (*insert_vm_struct)(struct mm_struct *mm, struct vm_area_struct *vma) = (INSERT_VM_STRUCT_ADDR != 0) ?
+        (int (*)(struct mm_struct *mm, struct vm_area_struct *vma)) (INSERT_VM_STRUCT_ADDR + (long long)(((void *)kfree) - (void *)KFREE_ADDR)) : NULL;
+
+	if (!vm_area_alloc) {
+		LOG_ERROR(-ENOTSUPP, "vm_area_alloc() was not found");
+		return -ENOTSUPP;
+	}
+
+	if (!vm_area_free) {
+		LOG_ERROR(-ENOTSUPP, "vm_area_free() was not found");
+		return -ENOTSUPP;
+	}
+
+	if (!insert_vm_struct) {
+		LOG_ERROR(-ENOTSUPP, "insert_vm_struct() was not found");
+		return -ENOTSUPP;
+	}
+
+	fiemap = NULL;
+	task = get_current();
+
+	LOG_DEBUG("Getting cow file extents from filp=%p", filp);
+	LOG_DEBUG("Attempting page stealing from %s", get_task_comm(parent_process_name, task));
+
+	start_addr = get_unmapped_area(NULL, 0, PAGE_SIZE, 0, VM_READ | VM_WRITE);
+	if (IS_ERR_VALUE(start_addr))
+		return start_addr; // returns -EPERM if failed
+
+	elastio_snap_mm_lock(task->mm);
+
+	// we brutally impose a page to the parent process
+	vma = vm_area_alloc(task->mm);
+	if (!vma) {
+		ret = -ENOMEM;
+		LOG_ERROR(ret, "vm_area_alloc() failed");
+		elastio_snap_mm_unlock(task->mm);
+		return ret;
+	}
```

Comment:
Does it work fine on all supported kernels? :)

Reply:
I've moved `elastio_snap_mm_unlock()` to a separate function because it may change on different kernels. `vm_area_alloc()` was introduced starting from 4.17, and we'll need to do something about it

Reply:
Let's add `TODO` and create an issue in the epic for it.

Reply:
Done